### PR TITLE
[QC] Add FP8 W8A8 variable-length FlashAttention-2

### PR DESCRIPTION
**Pull request by ggbondbest**
### PR Category

Operator | OP Test | Benchmark

### Type of Change

New Feature | Performance Optimization

### Description

Add `flaggems_vllm.flash_attn_varlen_func_w8a8_fp8` for variable-length
FlashAttention-2 with block-wise FP8 Q/K/V, per-128-token-block FP32 descales,
and FP16/BF16 output on NVIDIA Hopper or newer.

- Support packed uniform and ragged sequences, paged KV cache, causal and
  non-causal attention, and optional LSE output.
- Match the BF16 varlen head-dimension contract: every multiple of 8 from 8
  through 256, including non-power-of-two dimensions.
- Preserve the public `flash_attn_varlen_func` signature.
- Retain the optimized kernel path from flagos-ai/FlagGems#3877 for uniform
  packed D64/D128 inputs and extend the variable-length path to all dimensions.
- Zero-pad masked dimension lanes inside the kernel, with a minimum FP8 QK
  reduction tile of 32. Keep tensor shapes and the softmax scale unchanged.
- Split wide packed output dimensions into D64 slices to limit accumulator
  register pressure while retaining full-width QK computation.
- Compile short/long KV loops separately so precision correction does not
  introduce a runtime branch into the long-sequence PV loop. Mixed batches
  partition CTAs between the two variants by effective KV length.
- Use E4M3 probabilities even with E5M2 Q/K/V. Add an FP8 residual PV dot only
  for short KV sequences and causal/local boundary tiles that need additional
  rounding accuracy; this is not a BF16 compute fallback.
- Add dedicated accuracy and benchmark files named
  `test_flash_attn_varlen_func_w8a8_fp8.py` under `tests/` and `benchmark/`.
  The original `test_flash_attn_varlen_func.py` files are unchanged.
- Register/export the operator and use standard filename-based CI selection.
- Compare accuracy and performance with
  `vllm.vllm_flash_attn.flash_attn_interface.flash_attn_varlen_func`, explicitly
  selecting `fa_version=2` and BF16 reference Q/K/V.

The supported inference path requires equal query and KV head counts and
rejects dropout, context parallelism, and split-KV API mode before dispatch.
Production code does not use a Torch or BF16 compute fallback.

Autotune exemption: the uniform packed fast path retains the validated
`triton.autotune` candidates migrated from #3877. The variable-length path
continues to use the existing `mha_block_16/32/64/128` heuristics, Hopper shape
overrides, and a bounded wide-dimension configuration; this compatibility
extension does not introduce a new autotuner.

### Issue

- Kernel optimization reference: flagos-ai/FlagGems#3877
- Operator migration: flagos-ai/FlagGems#5817

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Validation

- H20, PyTorch 2.11.0+cu128, vLLM 0.20.2.
- Accuracy: 511 cases passed across four shards, including all 32 supported
  head dimensions, FP16/BF16 output, ragged sequences, paged KV cache, and LSE
  checks.
- Include unit-scale random inputs, non-contiguous row/head layouts, long
  sequences, and pure-short KV boundaries (1 and 128 tokens).
- Accuracy reference inputs are dequantized from the same FP8 Q/K/V tensors
  directly to BF16. Existing tolerances are unchanged: output `rtol=1e-2`,
  `atol=2e-2`; LSE `rtol=1e-2`, `atol=5e-2`.
- Test/benchmark Q/K preparation uses block-diagonal orthogonal Hadamard
  transforms for non-power-of-two dimensions without changing their sizes.
- Performance uses CUDA Graph replay for both operators. Quantization and
  BF16 conversion happen before timing; both operators reuse output buffers.
- The benchmark's dtype field describes the W8A8 source/output dtype. The
  vLLM baseline remains BF16 for both dtype groups.
- Full-repository pre-commit and CI pin checks pass; 54 CI helper tests and
  16 subtests passed.

```shell
python -m pytest -q tests/test_flash_attn_varlen_func_w8a8_fp8.py
python -m pytest -q -s benchmark/test_flash_attn_varlen_func_w8a8_fp8.py \
  --mode cudagraph --level comprehensive
python -m unittest discover -s tools -p 'test_ci_*.py'
python tools/check_ci_pins.py
pre-commit run --all-files --show-diff-on-failure
```

### Performance

The comprehensive benchmark covers 178 shapes for each W8A8 output dtype
(356 measurements), retaining all 54 original shapes and adding an all-dimension
sweep with uniform/ragged and causal/non-causal cases plus long D192/D256
sequences. The core set contains 10 shapes. Speedup is baseline latency divided
by W8A8 latency.

All 356 measurements completed successfully on H20 with CUDA Graph replay
(`--iter 100`, median mode). Both output dtype groups cover all 32 dimensions.

Across all cases, speedup ranges from 0.647x to 2.383x;
geometric mean is 1.325x for FP16 output and 1.323x for
BF16 output.

This is complete dimension support, not a guarantee of acceleration for every
shape: 62 of 356 measurements are below 1x (48 below 0.9x). Some unaligned
head dimensions and small causal ragged workloads remain slower. The original
54 shapes have geometric means of 1.629x/1.627x for FP16/BF16 output, with one
long D128 causal case per dtype slightly below baseline (about 0.965x).
The added 124 shapes have geometric means of 1.211x/1.209x respectively. No accuracy tolerance
was relaxed to improve performance.

These fixed representative rows use BF16 W8A8 output and non-causal attention.
The B=4 case uses Q lengths (17, 63, 129, 511) and KV lengths
(33, 1, 257, 513); all B=1 cases are uniform.

B | Max Q / KV | H / D | vLLM FA2 BF16 (ms) | W8A8 FP8 (ms) | Speedup
-- | -- | -- | -- | -- | --
1 | 512 / 512 | 8 / 8 | 0.01428 | 0.01270 | 1.125x
1 | 512 / 512 | 8 / 24 | 0.01461 | 0.01278 | 1.144x
1 | 512 / 512 | 32 / 64 | 0.03502 | 0.01887 | 1.856x
1 | 2048 / 2048 | 32 / 64 | 0.43355 | 0.21981 | 1.972x
1 | 512 / 512 | 8 / 96 | 0.02802 | 0.01672 | 1.676x
1 | 512 / 512 | 16 / 128 | 0.03546 | 0.02218 | 1.599x
1 | 512 / 512 | 8 / 192 | 0.04789 | 0.02771 | 1.728x
4 | 511 / 513 | 8 / 192 | 0.05390 | 0.04374 | 1.232x
1 | 4096 / 4096 | 8 / 192 | 1.42492 | 1.15513 | 1.234x
1 | 4096 / 4096 | 8 / 256 | 1.63670 | 1.54537 | 1.059x

### CODE REVIEW

File: tests/test_flash_attn_varlen_func.py [hunk 1]
```diff
@@ -12,18 +12,35 @@
...
+import math
 from typing import List, Optional, Tuple
 
 import pytest
 import torch
+import triton
 
 import flaggems_vllm
+from flaggems_vllm.ops import flash_attn_varlen_func_w8a8_fp8 as w8a8_varlen
+from flaggems_vllm.ops.attention import flash_attn_varlen_func as fa2_varlen
```
Comment (0x45f) [inline tests/test_flash_attn_varlen_func.py:25]: now we compare result with torch，can we add uts and compare result with flaggems_vllm.flash_attn_varlen_func and vllm.flash_attn_varlen_func

Comment (ggbondbest) [reply tests/test_flash_attn_varlen_func.py]: Since this is only an accuracy test, using PyTorch as the reference should be sufficient? The performance benchmark already compares against flaggems_vllm.flash_attn_varlen_func

Comment (ggbondbest) [reply tests/test_flash_attn_varlen_func.py:25]: already fixed. compare with flaggems_vllm.flash_attn_varlen_func

**Comment by 0x45f** [comment]
please fix conflict and code-style

File: tests/test_flash_attn_varlen_func.py [hunk 2]
```diff
@@ -370,3 +387,561 @@ def test_flash_attn_varlen_func_swap_qg(
...
+
+
+def _assert_w8a8_attention_close(actual, expected):
+    torch.testing.assert_close(
+        actual.float(), expected.float(), rtol=1.0e-2, atol=2.0e-2
+    )
+
+
+@pytest.mark.flash_attn_varlen_func_w8a8_fp8
+def test_flash_attn_varlen_func_w8a8_fp8_signature():
```
Comment (0x45f) [inline tests/test_flash_attn_varlen_func.py]: add new test file `test/test_flash_attn_varlen_func_w8a8_fp8.py`

**Review by 0x45f — COMMENTED** [review]
I want to know whether `vllm.flash_attn_varlen_func` supports `w8a8_fp8` input tensors, and whether its behavior is consistent with that in gems

File: tests/test_flash_attn_varlen_func.py (hunk 2 above)
Comment (ggbondbest) [reply tests/test_flash_attn_varlen_func.py]: it seems that benchmark performance testing requires creating new files;precision testing,however,doesn't seem to require this?  and vllm fa2 doesn't support fp8  only bf16/fp16

Comment (0x45f) [reply tests/test_flash_attn_varlen_func.py]: please move precision  uts to `test/test_flash_attn_varlen_func_w8a8_fp8.py`

Comment (ggbondbest) [reply tests/test_flash_attn_varlen_func.py]: already moved

**Comment by 0x45f** [comment]
I think the baseline for performance testing should be the `vllm.flash_attn_varlen_func`

**Comment by ggbondbest** [comment]
This is an operator from the FlagGems library and belongs to the ATen operator category. So we only need to compare it against the Triton BF16 implementation or PyTorch? And only operators in the flaggems-vllm library need to be compared with vLLM at the same precision?

> I think the baseline for performance testing should be the `vllm.flash_attn_varlen_func`

**Comment by ggbondbest** [comment]
already compare with vllm. plz check again

File: src/flaggems_vllm/ops/flash_attn_varlen_func_w8a8_fp8.py [hunk 3]
```diff
@@ -0,0 +1,3427 @@
...
+    if q.dtype not in _FP8_DTYPES or k.dtype != q.dtype or v.dtype != q.dtype:
+        raise TypeError("q, k, and v must have the same supported FP8 dtype")
+    if q.device != k.device or q.device != v.device:
+        raise ValueError("q, k, and v must be on the same device")
+    if q.stride(-1) != 1 or k.stride(-1) != 1 or v.stride(-1) != 1:
+        raise NotImplementedError("q, k, and v must be contiguous in head_dim")
+    if k.shape != v.shape:
+        raise ValueError("k and v must have identical shapes")
+    if q.shape[-1] not in (64, 128) or k.shape[-1] != q.shape[-1]:
+        raise NotImplementedError("only head_dim 64 and 128 are supported")
```
Comment (0x45f) [inline src/flaggems_vllm/ops/flash_attn_varlen_func_w8a8_fp8.py]: why only support 64/128 head_dim

Comment (ggbondbest) [reply src/flaggems_vllm/ops/flash_attn_varlen_func_w8a8_fp8.py]: already adjusted. now support 8-256 head-dim

**Review by 0x45f — APPROVED** [review]

**Review by huangyiqun — APPROVED** [review]
